### ISSUE TITLE
NVMe Disk Support

### ISSUE THREAD

**Issue by SteVwonder** [open-mpi/hwloc#399]
### What version of hwloc are you using?
1.11.8, 1.11.11, 2.1.0

1.11.8 is coming from RHEL:
```
❯ rpm -qa '*hwloc*'                                                                                                                                                                                                                                                                       18:54:23 ()
hwloc-devel-1.11.8-4.el7.x86_64
hwloc-1.11.8-4.el7.x86_64
hwloc-gui-1.11.8-4.el7.x86_64
hwloc-libs-1.11.8-4.el7.x86_64
hwloc-2.1-0-0.ch6.x86_64
```
and a hand-compiled version from git: e6559c78

I also hand compiled a version of 1.11.11 from a git checkout of the tag:  38035a04
2.1.0 is coming from spack: `spack install hwloc@2.1`

### Which operating system and hardware are you running on?

```
❯ uname -a
Linux corona82 3.10.0-1127.0.0.1chaos.ch6.x86_64 #1 SMP Fri Apr 3 08:56:52 PDT 2020 x86_64 x86_64 x86_64 GNU/Linux
```

### Details of the problem

We are using hwloc in Flux to auto-detect on-node resources available for scheduling.  Right now, we have it working for CPUs, GPUs, and memory.  We want to add node-local disk support as well.  Running `lstopo --whole-io --whole-system | grep -i -e disk -e nvme` with both 1.11.8 versions on a node with an NVMe drive turns up nothing.  Running the 1.11.11 hand-compiled version turns up:

```
            Block(Disk) L#0 "nvme0n1"
```

and the spack-built 2.1.0 version turns up:

```
          PCI 01:00.0 (NVMExp)
            Block(Disk) "nvme0n1"
  Block(Disk) "sda"
```

Switching to the xml format, we get some extra metadata from the 1.11.11 and 2.1.0 versions:

```
❯ ./lstopo --of xml | grep -A 10 -i -e nvme
            <object type="PCIDev" os_index="4096" name="HGST, Inc. Ultrastar SN200 Series NVMe SSD" pci_busid="0000:01:00.0" pci_type="0108 [1c58:0023] [1c58:0023] 02" pci_link_speed="3.938462">
              <info name="PCIVendor" value="HGST, Inc."/>
              <info name="PCIDevice" value="Ultrastar SN200 Series NVMe SSD"/>
              <info name="PCISlot" value="1"/>
              <object type="OSDev" name="nvme0n1" osdev_type="0">
                <info name="LinuxDeviceID" value="259:0"/>
                <info name="Model" value="HUSMR7616BDP301"/>
                <info name="SerialNumber" value="SDM000023759"/>
                <info name="Type" value="Disk"/>
              </object>
            </object>
          </object>
```

2.1.0:
```
❯ lstopo --of xml | grep -A 10 -i -e nvme
              <info name="PCIDevice" value="Ultrastar SN200 Series NVMe SSD"/>
              <object type="OSDev" gp_index="384" name="nvme0n1" subtype="Disk" osdev_type="0">
                <info name="Size" value="1562813784"/>
                <info name="SectorSize" value="4096"/>
                <info name="LinuxDeviceID" value="259:0"/>
                <info name="Model" value="HUSMR7616BDP301"/>
                <info name="SerialNumber" value="SDM000023759"/>
              </object>
            </object>
          </object>
```

My main question is: which version of hwloc should we be using such that we can see there is an NVMe device and the size of it.  Is that possible in the 1.11 series, and I'm just missing the right configure/runtime flag? Or is this a feature only available in the 2.X series of hwloc?  Diving into the code, it seems like a lot of the block device topology code is present in both 1.11 and 2.1, so I am wondering if this is a simple user mistake that I'm not seeing the size in 1.11.11 and not seeing the device at all in 1.11.8.

**Comment by bgoglin** [comment]
Hello Stephen

In 1.11.10, there's a fix related to Linux kernel 4.0 changing the way devices are exposed in sysfs (block class -> generic class).  It looks like Redhat backported some of those changes in their 3.10 kernel (tested on CentOS 7 and kernel 3.10.0-957.el7.x86_64). It might explain why you're not seeing anything with 1.11.8.

Regarding the size, it was added in master when we started the big 2.0 rework after branching 1.11 in 2015. I was planning to backport it but the appropriate attributes weren't decided yet, so I waited and forgot about it, sorry :/ So looks you'll need 2.0 here. At least my git branch 1.11 still doesn't show the size.

Brice
